src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1064,9 +1064,49 @@ BoundDecisionDagNode finalState(SyntaxNode syntax, LabelSymbol label, ImmutableA
                     fromTestPassing = fromTestPassing.Intersect(tempValuesBeforeTest);
                     fromTestFailing = fromTestFailing.Intersect(tempValuesBeforeTest);
                 }
+                bool truePossible = !fromTestPassing.IsEmpty;
+                bool falsePossible = !fromTestFailing.IsEmpty;
+                if (relation == BinaryOperatorKind.Equal &&
+                    input.Source is BoundDagPropertyEvaluation { IsLengthOrCount: true } s)
+                {
+                    Debug.Assert(value.Discriminator == ConstantValueTypeDiscriminator.Int32);
+                    int lengthValue = value.Int32Value;
+                    int commonIndex = -1; // common index from start, -1 means we don't have any
+                    IValueSet? commonValues = null;
+                    var temps = ArrayBuilder<BoundDagTemp>.GetInstance();
+                    foreach (var (temp, remainingValues) in values)
+                    {
+                        if (temp.Source is BoundDagIndexerEvaluation e)
```

Comment:
🚧 This probably won't work for indirect tests. e.g. `[{X: p, Y: q}, ..]`.

Reply:
Ah, good point. Need to think about that scenario... Let's add some tests to illustrate the issue.

Reply:
Wouldn't the same approach work? 
We can treat the temp from `[0].X` as equivalent to that from `[^1].X` and update their remaining values.
This could go even deeper `[0].X.[0].Y` is equivalent to `[^1].X.[0].Y` if `.Length` is `.

Reply:
Yes, we should walk up the input chain just as we do in CheckConsistentDecision. I want to see if we could add conditionally dependent inputs right when we found them so we don't need to do the same here.